### PR TITLE
FOGL-2621 plugin discovery system tests added

### PR DESCRIPTION
**Pull request by ashish-jabble**
Verified on below environments: 
- [x] ub16.04
- [x] raspbian

**O/P**

```
$ pytest -s -vv test_plugin_discovery.py
============================================================= test session starts =============================================================
platform linux -- Python 3.5.3, pytest-3.6.0, py-1.8.0, pluggy-0.6.0 -- /usr/local/bin/python3.5
cachedir: ../.pytest_cache
rootdir: /home/foglamp/Development/systemTests/FogLAMP/tests/system/python, inifile: pytest.ini
plugins: mock-1.10.0, cov-2.5.1, asyncio-0.8.0, allure-adaptor-1.7.10, aiohttp-0.3.0
collected 15 items                                                                                                                            

../TestPluginDiscovery test_cleanup <- api/test_plugin_discovery.py FogLAMP killed.
This script will remove all data stored in the SQLite3 datafile '/home/foglamp/Development/systemTests/FogLAMP/data/foglamp.db'
Enter YES if you want to continue: Starting FogLAMP v1.5.1......
FogLAMP started.
PASSED
../TestPluginDiscovery test_default_all_plugins_installed <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_all_plugins_installed <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_all_plugins_installed <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_default_plugins_installed_by_type <- api/test_plugin_discovery.py PASSED
../TestPluginDiscovery test_south_plugins_installed <- api/test_plugin_discovery.py Cloning into '/tmp/foglamp-south-sinusoid'...
remote: Enumerating objects: 59, done.
remote: Counting objects: 100% (59/59), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 266 (delta 21), reused 36 (delta 15), pack-reused 207
Receiving objects: 100% (266/266), 46.77 KiB | 0 bytes/s, done.
Resolving deltas: 100% (83/83), done.
Checking connectivity... done.
No such external dependency needed for sinusoid plugin.
sinusoid plugin is installed.
Cloning into '/tmp/foglamp-south-random'...
remote: Enumerating objects: 58, done.
remote: Counting objects: 100% (58/58), done.
remote: Compressing objects: 100% (44/44), done.
remote: Total 158 (delta 20), reused 28 (delta 8), pack-reused 100
Receiving objects: 100% (158/158), 49.48 KiB | 0 bytes/s, done.
Resolving deltas: 100% (67/67), done.
Checking connectivity... done.
No external dependency needed for random plugin.
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No options set.
   +Using found FOGLAMP_ROOT /home/foglamp/Development/systemTests/FogLAMP
-- Using FOGLAMP_ROOT for includes
   +/home/foglamp/Development/systemTests/FogLAMP/C/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/services/common/include
-- Using FOGLAMP_ROOT for libs 
   +/home/foglamp/Development/systemTests/FogLAMP/cmake_build/C/lib
-- Using third-party includes /home/foglamp/Development/systemTests/FogLAMP/C/thirdparty
-- Installing Random in /home/foglamp/Development/systemTests/FogLAMP/plugins/south/Random
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/foglamp-south-random/build
[ 25%] Generating version header
Scanning dependencies of target Random
[ 50%] Building CXX object CMakeFiles/Random.dir/plugin.o
[ 75%] Building CXX object CMakeFiles/Random.dir/random.o
[100%] Linking CXX shared library libRandom.so
[100%] Built target Random
[100%] Built target Random
Install the project...
-- Install configuration: ""
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/south/Random/libRandom.so.1
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/south/Random/libRandom.so
-- Set runtime path of "/home/foglamp/Development/systemTests/FogLAMP/plugins/south/Random/libRandom.so.1" to ""
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
random plugin is installed.
PASSED
../TestPluginDiscovery test_north_plugins_installed <- api/test_plugin_discovery.py Cloning into '/tmp/foglamp-north-http'...
remote: Enumerating objects: 35, done.
remote: Counting objects: 100% (35/35), done.
remote: Compressing objects: 100% (19/19), done.
remote: Total 254 (delta 5), reused 25 (delta 4), pack-reused 219
Receiving objects: 100% (254/254), 50.41 KiB | 0 bytes/s, done.
Resolving deltas: 100% (62/62), done.
Checking connectivity... done.
No such external dependency needed for http plugin.
http plugin is installed.
Cloning into '/tmp/foglamp-north-thingspeak'...
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (30/30), done.
remote: Total 92 (delta 13), reused 22 (delta 6), pack-reused 50
Unpacking objects: 100% (92/92), done.
Checking connectivity... done.
No external dependency needed for thingspeak plugin.
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No options set.
   +Using found FOGLAMP_ROOT /home/foglamp/Development/systemTests/FogLAMP
-- Using FOGLAMP_ROOT for includes
   +/home/foglamp/Development/systemTests/FogLAMP/C/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/services/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/plugins/filter/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/rapidjson/include
-- Using FOGLAMP_ROOT for libs 
   +/home/foglamp/Development/systemTests/FogLAMP/cmake_build/C/lib
-- Using third-party includes /home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/Simple-Web-Server
-- Installing ThingSpeak in /home/foglamp/Development/systemTests/FogLAMP/plugins/north/ThingSpeak
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/foglamp-north-thingspeak/build
[ 25%] Generating version header
Scanning dependencies of target ThingSpeak
[ 50%] Building CXX object CMakeFiles/ThingSpeak.dir/ThingSpeak.o
[ 75%] Building CXX object CMakeFiles/ThingSpeak.dir/plugin.o
[100%] Linking CXX shared library libThingSpeak.so
[100%] Built target ThingSpeak
[100%] Built target ThingSpeak
Install the project...
-- Install configuration: ""
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/north/ThingSpeak/libThingSpeak.so.1
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/north/ThingSpeak/libThingSpeak.so
-- Set runtime path of "/home/foglamp/Development/systemTests/FogLAMP/plugins/north/ThingSpeak/libThingSpeak.so.1" to ""
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
thingspeak plugin is installed.
PASSED
../TestPluginDiscovery test_filter_plugins_installed <- api/test_plugin_discovery.py Cloning into '/tmp/foglamp-filter-rms'...
remote: Enumerating objects: 148, done.
remote: Counting objects: 100% (148/148), done.
remote: Compressing objects: 100% (76/76), done.
remote: Total 148 (delta 65), reused 112 (delta 42), pack-reused 0
Receiving objects: 100% (148/148), 40.81 KiB | 0 bytes/s, done.
Resolving deltas: 100% (65/65), done.
Checking connectivity... done.
No external dependency needed for rms plugin.
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No options set.
   +Using found FOGLAMP_ROOT /home/foglamp/Development/systemTests/FogLAMP
-- Using FOGLAMP_ROOT for includes
   +/home/foglamp/Development/systemTests/FogLAMP/C/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/services/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/plugins/filter/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/rapidjson/include
-- Using FOGLAMP_ROOT for libs 
   +/home/foglamp/Development/systemTests/FogLAMP/cmake_build/C/lib
-- Using third-party includes /home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/Simple-Web-Server
-- Installing rms in /home/foglamp/Development/systemTests/FogLAMP/plugins/filter/rms
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/foglamp-filter-rms/build
[ 25%] Generating version header
Scanning dependencies of target rms
[ 75%] Building CXX object CMakeFiles/rms.dir/plugin.cpp.o
[ 75%] Building CXX object CMakeFiles/rms.dir/rms.cpp.o
[100%] Linking CXX shared library librms.so
[100%] Built target rms
[100%] Built target rms
Install the project...
-- Install configuration: ""
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/filter/rms/librms.so.1
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/filter/rms/librms.so
-- Set runtime path of "/home/foglamp/Development/systemTests/FogLAMP/plugins/filter/rms/librms.so.1" to ""
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
rms plugin is installed.
PASSED
../TestPluginDiscovery test_delivery_plugins_installed <- api/test_plugin_discovery.py Cloning into '/tmp/foglamp-notify-slack'...
remote: Enumerating objects: 60, done.
remote: Counting objects: 100% (60/60), done.
remote: Compressing objects: 100% (31/31), done.
remote: Total 60 (delta 19), reused 46 (delta 13), pack-reused 0
Unpacking objects: 100% (60/60), done.
Checking connectivity... done.
[sudo] password for foglamp: 
Reading package lists... Done
Building dependency tree       
Reading state information... Done
libcurl4-openssl-dev is already the newest version (7.47.0-1ubuntu2.12).
The following packages were automatically installed and are no longer required:
  libllvm5.0 linux-headers-4.13.0-41 linux-headers-4.13.0-41-generic linux-headers-4.13.0-43 linux-headers-4.13.0-43-generic
  linux-headers-4.13.0-45 linux-headers-4.13.0-45-generic linux-headers-4.15.0-29 linux-headers-4.15.0-29-generic
  linux-image-4.13.0-41-generic linux-image-4.13.0-43-generic linux-image-4.13.0-45-generic linux-image-4.15.0-29-generic
  linux-image-extra-4.13.0-41-generic linux-image-extra-4.13.0-43-generic linux-image-extra-4.13.0-45-generic linux-modules-4.15.0-29-generic
  linux-modules-extra-4.15.0-29-generic
Use 'sudo apt autoremove' to remove them.
0 upgraded, 0 newly installed, 0 to remove and 37 not upgraded.
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No options set.
   +Using found FOGLAMP_ROOT /home/foglamp/Development/systemTests/FogLAMP
-- Using FOGLAMP_ROOT for includes
   +/home/foglamp/Development/systemTests/FogLAMP/C/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/services/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/plugins/filter/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/rapidjson/include
-- Using FOGLAMP_ROOT for libs 
   +/home/foglamp/Development/systemTests/FogLAMP/cmake_build/C/lib
-- Using third-party includes /home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/Simple-Web-Server
-- Installing slack in /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationDelivery/slack
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/foglamp-notify-slack/build
[ 25%] Generating version header
Scanning dependencies of target slack
[ 50%] Building CXX object CMakeFiles/slack.dir/slack.cpp.o
[ 75%] Building CXX object CMakeFiles/slack.dir/plugin.cpp.o
[100%] Linking CXX shared library libslack.so
[100%] Built target slack
[100%] Built target slack
Install the project...
-- Install configuration: ""
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationDelivery/slack/libslack.so.1
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationDelivery/slack/libslack.so
-- Set runtime path of "/home/foglamp/Development/systemTests/FogLAMP/plugins/notificationDelivery/slack/libslack.so.1" to ""
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
slack plugin is installed.
PASSED
../TestPluginDiscovery test_rule_plugins_installed <- api/test_plugin_discovery.py Cloning into '/tmp/foglamp-rule-outofbound'...
remote: Enumerating objects: 42, done.
remote: Counting objects: 100% (42/42), done.
remote: Compressing objects: 100% (24/24), done.
remote: Total 42 (delta 8), reused 34 (delta 7), pack-reused 0
Unpacking objects: 100% (42/42), done.
Checking connectivity... done.
No external dependency needed for outofbound plugin.
Cloning into '/tmp/foglamp-service-notification'...
remote: Enumerating objects: 79, done.
remote: Counting objects: 100% (79/79), done.
remote: Compressing objects: 100% (43/43), done.
remote: Total 659 (delta 26), reused 56 (delta 19), pack-reused 580
Receiving objects: 100% (659/659), 198.00 KiB | 236.00 KiB/s, done.
Resolving deltas: 100% (389/389), done.
Checking connectivity... done.
-- The C compiler identification is GNU 5.5.0
-- The CXX compiler identification is GNU 5.5.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- No options set.
   +Using found FOGLAMP_ROOT /home/foglamp/Development/systemTests/FogLAMP
-- Using FOGLAMP_ROOT for includes
   +/home/foglamp/Development/systemTests/FogLAMP/C/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/services/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/plugins/filter/common/include
   +/home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/rapidjson/include
-- Using FOGLAMP_ROOT for libs 
   +/home/foglamp/Development/systemTests/FogLAMP/cmake_build/C/lib
-- Notification server includes dir set to /tmp/foglamp-service-notification/C/services/common/include
-- Using third-party includes /home/foglamp/Development/systemTests/FogLAMP/C/thirdparty/Simple-Web-Server
-- Installing OutOfBound in /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationRule/OutOfBound
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/foglamp-rule-outofbound/build
[ 33%] Generating version header
Scanning dependencies of target OutOfBound
[ 66%] Building CXX object CMakeFiles/OutOfBound.dir/plugin.o
[100%] Linking CXX shared library libOutOfBound.so
[100%] Built target OutOfBound
[100%] Built target OutOfBound
Install the project...
-- Install configuration: ""
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationRule/OutOfBound/libOutOfBound.so.1
-- Installing: /home/foglamp/Development/systemTests/FogLAMP/plugins/notificationRule/OutOfBound/libOutOfBound.so
-- Set runtime path of "/home/foglamp/Development/systemTests/FogLAMP/plugins/notificationRule/OutOfBound/libOutOfBound.so.1" to ""
/home/foglamp/Development/systemTests/FogLAMP/tests/system/python/api
outofbound plugin is installed.
PASSED
====== 15 passed in 76.38 seconds ====
```

### CODE REVIEW

File: tests/system/python/api/test_plugin_discovery.py [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+        path = "$FOGLAMP_ROOT/tests/system/python/scripts/install_c_plugin {} {} {}".format(
+            branch, type, plugin)
+    try:
+        subprocess.run([path], shell=True, check=True)
+    except subprocess.CalledProcessError:
+        assert False, "{} plugin installation failed".format(plugin)
+
+    # Cleanup code
+    subprocess.run(["rm -rf /tmp/foglamp-{}-{}".format(type, plugin)], shell=True, check=True)
+    # TODO: FOGL-2642 - Remove notificationDelivery And notificationRule plugins
```
Comment (praveen-garg) [inline tests/system/python/api/test_plugin_discovery.py]: let's fix 2642 first.

File: tests/system/python/api/test_plugin_discovery.py [hunk 2]
```diff
@@ -0,0 +1,177 @@
...
+                    "egrep -v '(__init__.py)' | xargs rm -rf"], shell=True, check=True)
+    subprocess.run(["path=$FOGLAMP_ROOT/python/foglamp/plugins/north; "
+                    "find $path -maxdepth 1 | grep -v \"^${path}$\" | "
+                    "egrep -v '(pi_server|ocs|common|empty|README.rst|__init__.py)' | xargs rm -rf"],
+                   shell=True, check=True)
+    subprocess.run(["path=$FOGLAMP_ROOT/plugins/north; "
+                    "find $path -maxdepth 1 | grep -v \"^${path}$\" | "
+                    "egrep -v '(PI_Server_V2|ocs_V2)' | xargs rm -rf"],
+                   shell=True, check=True)
+    subprocess.run(["path=$FOGLAMP_ROOT/plugins/filter; "
```
Comment (praveen-garg) [inline tests/system/python/api/test_plugin_discovery.py]: let's have a script for this, and pass arg for south, north... if no arg passed, rm all

File: tests/system/python/api/test_plugin_discovery.py [hunk 3]
```diff
@@ -0,0 +1,177 @@
...
+class TestPluginDiscovery:
+
+    def test_default_all_plugins_installed(self, reset_plugins, reset_and_start_foglamp, foglamp_url):
+        conn = http.client.HTTPConnection(foglamp_url)
+        conn.request("GET", '/foglamp/plugins/installed')
+        r = conn.getresponse()
+        assert 200 == r.status
+        r = r.read().decode()
+        jdoc = json.loads(r)
+        assert 4 == len(jdoc['plugins'])
```
Comment (praveen-garg) [inline tests/system/python/api/test_plugin_discovery.py]: which 4? (2 north_c and 2 north python)? please put a comment

File: tests/system/python/api/test_plugin_discovery.py [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+        for plugin in jdoc['plugins']:
+            assert 'north' == plugin['type']
+            assert 'south' not in plugin['type']
+            assert 'filter' not in plugin['type']
+            assert 'notify' not in plugin['type']
+            assert 'rule' not in plugin['type']
```
Comment (praveen-garg) [inline tests/system/python/api/test_plugin_discovery.py]: I think there is still confusion, and hence FOGL-2642 for above two assertion

File: tests/system/python/api/test_plugin_discovery.py [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+            assert 'config' not in plugin
```
Comment (praveen-garg) [inline tests/system/python/api/test_plugin_discovery.py]: please put a newline and comment,  config is not expected by default. And repeat the same test with `?config=true`

File: tests/system/python/api/test_plugin_discovery.py (hunk 1 above)
Comment (ashish-jabble) [reply tests/system/python/api/test_plugin_discovery.py]: fyi, We are not BLOCKED with it as we can fetch notification plugins from `curl -sX GET http://localhost:8081/foglamp/notification/plugin | jq` And GUI is using the same endpoint.

Comment (ashish-jabble) [reply tests/system/python/api/test_plugin_discovery.py]: > let's fix 2642 first.

https://github.com/foglamp/FogLAMP/pull/1486

**Pull request merged**